### ISSUE TITLE
Treelist chaperone/contract problems with functional update

### ISSUE THREAD

**Issue by LiberalArtist** [racket/racket#4922]
(I found this while thinking further about https://github.com/racket/racket/pull/4913#discussion_r1472969940.)

Consider the following program:
```racket
#lang racket
(require racket/treelist)
(define/contract lazy-natural-treelist
  (treelist/c natural? #:lazy? #t #:flat? #f)
  (treelist 1 2 3))
(treelist-add lazy-natural-treelist "abc")
```

With the 8.12.0.7-2024-02-03-901a31301a snapshot, the implicit printing produces the following error:

```
2024-02-03-901a31301a/collects/racket/contract/private/blame.rkt:346:0: lazy-natural-treelist: broke its own contract
  promised: natural?
  produced: "abc"
  in: the elements of
      (treelist/c natural?)
  contract from: 
      (definition lazy-natural-treelist)
  blaming: (definition lazy-natural-treelist)
   (assuming the contract is correct)
  at: 3-unsaved-editor:3:17
```

First of all, even if this program ought to produce an error, it seems like the blame ought to lie with the party that added the prohibited value.

More broadly, this example illustrates some of what I found surprising about the design of `chaperone-treelist`.

I typically think about chaperones, especially in the context of contracts, as enabling interposition on operations that provide a "communication channel" across a "boundary". I don't think of immutable values with functional update as providing such a communication channel. While a library may have specific requirements for the immutable values I supply to its functions, I expect to be able to use the values it returns to me freely within my own code. For example, there were a variant of the `xml` library that used a treelist-based `xexpr/c`, I would find it very surprising if I couldn't take parts of treelists read from XML and `treelist-add`/`treelist-cons` non-`xexpr/c` values to them.

The current [implementation of `treelist/c`](https://github.com/racket/racket/blob/901a31301a49049a911be0b3d2128fedfa1a0014/racket/collects/racket/contract/private/treelist.rkt#L119-L129) only supplies a substantive `ref-proc` to `chaperone-treelist`, but I found this example while thinking about the way `set-proc`, `insert-proc`, and `append-proc` more directly interpose on functional update. The only analogous precedent I could find is `chaperone-hash`: it allows interposition on `hash-set` and other functional update procedures, seemingly by analogy to the imperative variants on mutable hashtables (which definitely do serve as communication channels), but it's not clear that the implications have been fully explored. In particular, the prose documentation for `impersonate-hash` talks about both the imperative and the functional variants, but only mutable hashtables can have non-chaperone impersonators. Using `chaperone-hash` can produce similar surprising behavior to the example above (though I note that there are no lazy contracts on hashtables, so the blame issue doesn't apply):

```racket
#lang racket
(define h
  (chaperone-hash
   #hash()
   (λ (h k) ; ref-proc
     (values k (λ (h k v)
                 v)))
   (λ (h k v) ; set-proc
     (unless (symbol? k)
       (error "not allowed"))
     (values k v))
   (λ (h k) ; remove-proc
     k)
   (λ (h k) ; key-proc
     k)))
(hash-set h 1 #\a)
```

Finally (and somewhat tangentially), [re](https://github.com/racket/racket/pull/4913#discussion_r1472969940):

> Yes, using an extra impersonator property to allow unwrapping is the usual way to go. The mutable-treelist implementation uses that trick itself, but it piggy-backs on the table of chaperone procedures that already uses a property.

I don't think this technique quite works with the current `chaperone-treelist`. As the following program illustrates, currently, functionally updating a chaperoned treelist drops impersonator properties from the result. For unwrapping via a property to work, the result would need not only to have the property, but to map it to the non-chaperoned version of the resulting treelist.

```racket
#lang racket
(require racket/treelist)
(define-values [i-p:orig has-orig? get-orig]
  (make-impersonator-property 'orig))
(define t0
  (treelist 1 2 3))
(define t0+chap
  (chaperone-treelist
   t0
   #f
   (λ (t i x)
     x)
   (λ (t i x)
     x)
   (λ (a b)
     b)
   i-p:orig t0))
(define t1
  (treelist-add t0+chap #\a))
(impersonator? t1) ;-> #t
(has-orig? t1) ;-> #f
```

**Comment by rfindler** [comment]
From the "what do I do in the contract system point of view", I guess my question is: is the printout that happens in this program avoidable?
```
#lang racket
(require racket/treelist)

(define t1 (treelist 1 2 3))
(define t2 (chaperone-treelist
            t1
            (λ (t i v) (printf "ref: @~s ~s\n" i v) v)
            (λ (t i v) v)
            (λ (t i v) v)
            (λ (t1 t2) t2)))
(define t3 (treelist-set t2 0 "one"))
(treelist-ref t3 0)
```

That is, when we access an index value that's crossed the boundary already is there a way to keep track of that and not use the chaperone for that index? Just from a little bit of thought, it seems unlikely that there is a good answer here! One possible route is that lazy contracts are just not supported (and we can add to the documentation to explain the algorithmic issue).

ISTR some fairly extension discussion at northwestern about this issue in the context of hashes with @chrdimo (and others who were at Northwestern at the time). I don't think that we ever made any changes to the contract system because of it, however, but maybe he has thoughts here.

**Comment by LiberalArtist** [comment]
> That is, when we access an index value that's crossed the boundary already is there a way to keep track of that and not use the chaperone for that index? Just from a little bit of thought, it seems unlikely that there is a good answer here!

If the problem with impersonator properties were fixed, one possibility might be to use a property to keep track of which indices you're interested in and do the necessary bookkeeping on set, insert, and append. (I think there would need to be a way to update properties on `treelist-sublist` and the take/drop operations, too, to keep them from  becoming O(n) via `treelist-ref`.) I don't know that this is a *good* answer: in particular, the overhead might be too high to be useful for most circumstances.

If we're assuming some changes to `chaperone-treelist`, possibly it could handle the bookkeeping for relevant indices directly, and it might be able to do so more efficiently than interposing on set etc. with an arbitrary procedure.

(I also noticed there isn't a `mutable-treelist/c` yet, but probably it makes sense to figure out the immutable ones first.)

**Comment by rfindler** [comment]
> (I also noticed there isn't a `mutable-treelist/c` yet, but probably it makes sense to figure out the immutable ones first.)

... I would have pushed it already, actually, but I just haven't found the time to sit down and write it.

**Comment by lexi-lambda** [comment]
I am a bit late to this party, but I stumbled into this today while trying to make sense of `treelist/c` and `chaperone-treelist`. A summary of my observations, discoveries, and thoughts:

* Matthew’s changes to `treelist/c` do avoid the issue illustrated by the program in the original post of this thread, so that is resolved, at least.

* I agree with @LiberalArtist that I find `chaperone-treelist` a bit odd. We do not have `chaperone-list`, after all.

  I do recall a discussion I had with @chrdimo in which he made the case that there is nothing fundamentally wrong with supporting interposition on immutable data (and we do permit `chaperone-vector` on immutable vectors), Racket just happens to not support it for lists. Moreover, differences in evaluation order between contracts on mutable and immutable data played a significant (but extremely subtle) role in the results of [Does Blame Shifting Work?](https://dl.acm.org/doi/pdf/10.1145/3371133). While researching this, I also found [Lazy Contract Checking for Immutable Data Structures](https://users.cs.northwestern.edu/~robby/publications/papers/ifl2007-fgr.pdf), which discusses precisely this question.

  However, `chaperone-treelist` is quite different from `chaperone-vector` and the `bounded-bst` contract in that paper in one key respect: it allows interposition on *functional updates*, not just accesses, which is what this issue is about. I like Philip’s characterization of why this is so surprising:

  > While a library may have specific requirements for the immutable values I supply to its functions, I expect to be able to use the values it returns to me freely within my own code. For example, there were a variant of the `xml` library that used a treelist-based `xexpr/c`, I would find it very surprising if I couldn't take parts of treelists read from XML and `treelist-add`/`treelist-cons` non-`xexpr/c` values to them.

  As it stands, `treelist/c` does not actually restrict functional updates, so this is currently only relevant if `chaperone-treelist` is used directly, which few people are likely to do. Still, I do find it surprising that it is possible at all.

* On paper, a lazy variant of `treelist/c` sounds good. If one has a treelist with tens of thousands of elements, it is hardly desirable to eagerly copy the entire thing and apply contract projections to all its elements every time it crosses a contract boundary, especially if only a couple elements will be needed. However, in practice, there are serious shortcomings of lazy treelist contracts that make me skeptical about their value:

  * The bookkeeping Matthew added to avoid the issue Philip highlighted eagerly allocates a new treelist of the same size as the input treelist to track which elements were from the original treelist. This means applying a lazy treelist contract still does exactly as much work as copying the entire input treelist, and the only thing that is skipped is applying contract projections to the elements!

  * Now, to be fair, if applying contract projections to the elements is expensive, that could still be a win. But this comes at a serious cost: the treelist *and all its descendants* are now permanently “tainted” by this contract wrapper, which has permanently increased the amount of memory it takes to represent the treelist twofold. And this is if the treelist passes a contract boundary just once!

    Each additional crossing of a contract boundary adds another bookkeeping wrapper which consumes time and space proportional to the size of the input, forever. This is not exponential, fortunately, but it’s still easy to see how quickly this could get out of hand.

    In theory, it seems like collapsible contracts could help substantially with this: it would require merging the bookkeeping structures used to track which elements are protected by the contracts, but that seems tractable. However, the blowup would still occur if collapsing isn’t possible for any reason. (And, either way, `treelist/c` currently is not collapsible.)

  None of this is in any way disclosed to users of `treelist/c`.

* It seems to me that a far less surprising design would be to dispense with the bookkeeping “shadow treelist” used to track which items belong to the original treelist and to instead have some way for the lazy projections to be attached to treelist nodes. If a node is ever touched, the projection could be forced for all 32 of its slots. Forcing a projection for a leaf node could apply projections to the elements of that node, while forcing a projection for a parent node could push the lazy projection down to its children.

  One nice thing about this design is that it would allow lazy `treelist/c` contracts to exist without `chaperone-treelist` having to provide any way to interpose upon anything other than reads. It also feels much closer to how a “lazy contract” on immutable data ought to work: it works the same way as an eager contract, but application of the projections is delayed until the structure is observed. (Perhaps this is just the Haskell programmer in me, but it would also be nice if forcing the application of the projections was memoized.)

* Regardless of all of the above, I do not understand *at all* how the default value for the `#:lazy?` argument to `treelist/c` is justified. Currently, it defaults to `#t` for chaperone contracts and to `#f` for both flat and impersonator contracts. Why?!

  If anything, it seems like laziness would be *more* desirable for flat contracts, since applying a flat projection can require doing far more work than applying a chaperone or impersonator projection (since the latter are generally lazier). As it stands, the lazy strategy is incurring the massive bookkeeping cost of maintaining the “shadow treelist” just to avoid applying a bunch of chaperone projections that are, most likely, just going to wrap each element in a wrapper and return.

  I do understand that `#:lazy?` cannot be `#t` for impersonator contracts because there is no `impersonate-treelist`. But defaulting it to `#t` at all seems like probably the wrong choice to me. I think it should just always default to `#f`.

**tl;dr**: I agree that `chaperone-treelist` seems weird, I think the current implementation of lazy `treelist/c` seems ill-advised, and I think it’s downright crazy that `#:lazy?` defaults to `#t` when `treelist/c` is given a chaperone contract.

**Comment by lexi-lambda** [comment]
A small correction to my previous comment: it seems like I was misreading the changelog, and Robby is to blame for the “shadow treelist” in `treelist/c`, not Matthew. I apologize to Matthew for any misdirected ire and to Robby for being the new victim of it. :)

**Comment by mfelleisen** [comment]
( @lexi-lambda Are you shifting blame? ) :->

**Comment by AlexKnauth** [comment]
> Forcing a projection for a leaf node could apply projections to the elements of that node

Would that have to mean accessing an element and checking its contract, would end up also checking the contracts of up to 31 of its neighboring elements?

Is there a way to avoid this? Would a way to avoid this have to include the bookkeeping that you're trying to avoid?

At least it seems better if the bookkeeping is lazy and applied only to leaf nodes that the user accesses, rather than eager for all elements whether accessed or not.

**Comment by lexi-lambda** [comment]
> Would that have to mean accessing an element and checking its contract, would end up also checking the contracts of up to 31 of its neighboring elements?

Yes, that is the scheme I had in mind. Personally, that doesn’t actually bother me at all—I see the laziness as purely a performance optimization, so I think some imprecision is fine—but perhaps other people feel otherwise.

**Comment by mflatt** [comment]
A clarification on the cost of `make-treelist` for a shadow treelist: it costs O(lg N) time, not O(N). So, it's not as slow as copying the whole treelist.

**Comment by lexi-lambda** [comment]
I thought about that, too. It would always be possible to use a plain old bytestring as a bitvector, and that could save on space by a factor of 32 or 64, depending on architecture. However, that would require the sort of unbounded copying a treelist is explicitly intended to avoid. Using a treelist of fixnums as a bitvector would be a more natural approach, which would accomplish essentially the same thing, albeit with slightly smaller factors.

Saving space by a factor of 61 is nothing to sneeze at, so it does seem like it could be a nice optimization, though you’d be making the usual time-space tradeoff as far as vectors of booleans are concerned. In this case, I think I’d probably take the space savings, since if you need maximum speed for your tight inner loop, you’re not going to be using lazy `treelist/c`. But I obviously haven’t profiled.

Either way, that optimization doesn’t change the asymptotics at all; it’s just a constant factor reduction (albeit a large one). In the worst case where every treelist node is functionally updated, you still pay linear space for each lazy `treelist/c` chaperone.
